### PR TITLE
fix(repl): add correction context after interruption

### LINKED ISSUE #1422: feat: inject correction context after user interruption (Escape)
## Problem

When a user interrupts an ongoing task (2× Escape) and corrects the model, the model ignores the correction and repeats the same wrong behavior.

**Observed on:** v0.15.0 with Qwen3.6 27B

**Reproduction:**
1. Give a task to the model
2. Model starts doing something unintended
3. Interrupt with 2× Escape
4. Type a correction ("don't do X, do Y instead")
5. Model restarts and does X again

## Why it happens

After interruption, the conversation history still contains the original instruction + already-executed tool calls. The model re-reads the context, sees it was doing X, and continues X — the correction arrives late in the context and is not weighted as a directive to change behavior.

## Proposed fix

After an interruption, inject a system-level signal before the user's correction message:

```
[INTERRUPTED] Previous attempt was stopped by the user.
Disregard prior tool calls. Follow only the next instruction.
```

This gives the model an explicit anchor to break momentum from the original task.

## Alternatives considered

- `/clear` — works but loses full history
- User rephrasing with emphasis — unreliable, puts burden on user

### PR DESCRIPTION
**Pull request by chioarub**
## Summary

- add a one-shot model-visible reminder after a user interrupts an active assistant turn
- keep slash commands, local UI commands, remote events, timeouts, and programmatic aborts from consuming or arming the reminder

## Impact

- user-facing impact: the next normal correction prompt is explicitly framed as corrective instead of letting the model continue the interrupted plan
- developer/maintainer impact: interruption state is session-scoped and tied to the exact model-bound query

## Testing

- [x] `bun run build`
- [x] `bun run smoke`
- [ ] `bun run check` — the feature branch and an untouched `upstream/main` baseline produced the same 43 existing failures; the branch added five passing tests in that comparison
- [x] `bun run typecheck`
- [x] focused tests: interruption state, prompt submission, REPL lifecycle wiring, and meta-message API behavior

## Notes

- provider/model path tested: provider-independent message construction; no live provider-backed REPL session
- screenshots attached: not applicable for this terminal interaction change
- follow-up work or known limitations: no rendered Ink/Escape integration test; the QueryGuard-backed lifecycle boundary and queued-input behavior are covered directly

Fixes #1422

## Summary by CodeRabbit

- **New Features**
  - Added one-time, session-scoped interruption-correction reminders for eligible *normal local* prompts, with full REPL turn lifecycle integration.
  - Introduced **request-only messages** to provide model context without compacting, persisting, exposing to tools, or being added to transcript state.
  - Added prompt submission control (`allowInterruptionCorrection`) and model-call lifecycle hooks (`onModelRequestStart`/`onModelRequestEnd`).

- **Bug Fixes**
  - Fixed interruption-correction arming/restoration behavior across cancellation, queued/concurrent turns, rewind/auto-restore, and session switching.
  - Improved provider-error normalization and feature-flag-dependent meta tagging.

- **Tests**
  - Expanded REPL, prompt submission/queue, interruption-correction, request-only propagation, and message normalization coverage.

### CODE REVIEW

**Review by jatmn — CHANGES_REQUESTED** [review]
---

## Medium — should address before merge

### M1. Reminder lost when the user cancels during the pre-query phase
**Files:** `src/utils/interruptionCorrection.ts:53-66` (`bindModelTurn`); `src/screens/REPL.tsx:3161-3176` (`onQuery`).

`bindModelTurn` sets `modelBoundQueryId` only when `!isAborted`, where `isAborted: abortController.signal.aborted` is captured **once**, at bind time — which sits *after* two `await` points (`mrOnBeforeQuery` at REPL.tsx:3162 and `onBeforeQueryCallback` at :3167) but *before* `onQueryImpl` (:3177).

If the user presses Escape during those pre-query awaits, `onCancel → handleCancellation` runs while `modelBoundQueryId` is still `null`. `shouldMarkInterruptionCorrection` requires a non-null `modelBoundQueryId`, so nothing arms. The abort then flips `signal.aborted`, and the later `bindModelTurn` skips setting the binding. **Result:** a legitimate user-initiated, local cancellation of an "active turn" produces no correction reminder — contradicting the stated feature.

This is a timing window (not the common case), but it is exactly the kind of cancellation the feature targets, so it should be closed.

**Suggested fix:** Either (a) move the binding decision to the point the model turn actually starts (inside `onQueryImpl`, after the pre-query awaits), or (b) drop the `!isAborted` guard from `bindModelTurn` and let `finishModelTurn`/`handleCancellation` do all clearing — an already-aborted turn that never bound cannot arm because `handleCancellation` runs *before* bind in that window and `modelBoundQueryId` is `null`. Add a test: "cancel during pre-query phase still arms the reminder."

### M2. Vacuous session-isolation assertion in `REPL.queryLifecycle.test.ts`
**File:** `src/screens/REPL.queryLifecycle.test.ts:126-127` (test `executes correction arming and consumption through QueryGuard`).

The test consumes the reminder at lines 120-124 (`takeReminder()` → reminder, then `takeReminder()` → `null`). It then switches `sessionId = 'session-b'` and asserts `takeReminder()` is `null` (line 127). But `pendingSessionId` is already `null` from the prior consumption, so that assertion passes **regardless of session matching** — it never actually exercises session scoping.

Reproduced against the real module: after consuming, switching the session and calling `takeReminder()` returns `null` (true even if session matching were broken).

Note: production session scoping *is* correctly covered by `interruptionCorrection.test.ts:55-57`, so behavior is fine — but this REPL-level assertion is misleading and should be fixed. This matches an existing CodeRabbit comment.

**Suggested fix:** Re-arm the tracker (start another model turn, `handleCancellation`, `forceEnd`) *before* switching to `session-b`, then assert `takeReminder()` is `null`. Per CodeRabbit's proposed diff.

### M3. `isNormalLocalUserPrompt` depends on undocumented invariants for `workload` / `agentId`
**File:** `src/utils/handlePromptSubmit.ts:43-56`.

The predicate correctly rejects slash (value *and* preExpansionValue), non-string `value`, undefined `preExpansionValue`, `skipSlashCommands`, `bridgeOrigin`, `isMeta`, `origin`, and `slashCommandOverride`. **It does not inspect `workload` or `agentId`**, which denote programmatic (cron / subagent) traffic on `QueuedCommand`.

Today this is safe **only** because of out-of-band invariants: (a) cron prompts set `isMeta: true`, and (b) subagent prompts are drained to a separate queue before reaching this code. Both are implicit. Any future programmatic enqueue that sets `preExpansionValue` (or forgets `isMeta`) would silently satisfy `isNormalLocalUserPrompt` and consume the one-shot meant for the next *human* prompt — violating the PR's own rule that programmatic prompts must not consume the reminder.

**Suggested fix (defense in depth):** add explicit
`command.workload === undefined && command.agentId === undefined`
to the predicate, and add test cases for both. This makes the "human-only" contract self-documenting and removes the coupling to queue-drain/snippet conventions.

---

## Low / Info — fast-follows

### L1. Reminder is consumed before `processUserInput`; lost if it throws
**File:** `src/utils/handlePromptSubmit.ts:507-535`.

`takeInterruptionCorrectionReminder?.()` is invoked at the top of the loop (clears `pendingSessionId`), but the reminder is only pushed into `newMessages` *after* the `await processUserInput` (line 533). If `processUserInput` throws for that command, the one-shot is consumed-but-never-shown. Severity is low (the turn fails anyway), but recommend deferring consumption until after a successful push, or re-arming on failure. Add a test.

### L2. `handlePromptSubmit` "interrupt on submit" never arms
**File:** `src/utils/handlePromptSubmit.ts:347-358` (`params.abortController?.abort('interrupt')`).

Submitting a *non-prompt* input mid-generation aborts the active turn **without** calling `onCancel`, so no reminder arms. This is consistent with the PR's intent to exclude slash/UI commands, but it is still a user-initiated interruption. Recommend a short comment documenting the deliberate exclusion (and a decision on whether non-prompt user interruptions that abort the model should arm).

### L3. `mergeUserMessages` strips `isMeta` from the reminder
**File:** `src/utils/messages.ts:1856-1907`.

When the reminder user-message is adjacent to the correction user-message, `mergeUserMessages` concatenates them; the non-`HISTORY_SNIP` path returns an object with no `isMeta` (only the snip branch sets it, and only when both operands are meta). Result: the reminder text is merged into a normal user turn and loses its `isMeta` identity. Functionally the text still reaches the model, but any meta-aware handling is lost. Consider keeping the reminder as a separate, non-mergeable message, or confirm this is acceptable.

### L4. Reminder persists in the transcript; verify `--resume` replay
**Files:** `src/utils/handlePromptSubmit.ts` (newMessages → onQuery → persistent history).

The `isMeta` reminder is written into the conversation history. On a resumed session, a stale "interrupted" signal could mislead the model. Not confirmed as a bug — needs author verification of serialization/resume behavior (whether `isMeta` user messages are re-sent and how the model treats a repeated interruption notice).

### L5. `takeReminder` silently destroys a session-mismatched reminder
**File:** `src/utils/interruptionCorrection.ts:100-107` (`consumeInterruptionCorrectionReminder`).

On a session change between arm and take, the reminder is consumed and discarded rather than deferred. Per the "session-scoped" design this may be acceptable, but it is an irreversible silent drop. Document the drop, or return the original `pendingSessionId` on mismatch so the reminder survives a transient session-id change.

### L6. `handleCancellation` re-reads the live guard
**File:** `src/screens/REPL.tsx:2322-2326`.

`onCancel` already captures `cancelContext` (line 2322) but `handleCancellation` re-reads `queryGuard.activeContext?.queryId`. Safe today because `forceEnd` runs after (line 2336). For robustness, pass the captured `cancelContext.queryId` into `handleCancellation` instead of re-reading.

### L7. Tracker trusts two distinct query-id sources
**File:** `src/utils/interruptionCorrection.ts:62-64` vs `:75-89`.

`bindModelTurn` stores `queryContext.queryId` while `handleCancellation` compares against `queryGuard.activeContext?.queryId`. Equal today by construction; recommend asserting/centralizing the single source of truth to prevent future drift.

**Comment by chioarub** [comment]
Thanks for the detailed review. The three medium items are fixed in `236a90c0`; the lower-severity observations were also checked against the current lifecycle and message contracts.

### M1. Reminder lost during the pre-query phase

**Fixed** in `236a90c0`. `InterruptionCorrectionTracker.runModelTurn` now binds the active model turn synchronously before either pre-query await and clears that exact query binding in `finally`; `REPL.tsx` wraps both pre-query callbacks and `onQueryImpl` in that boundary. `REPL.queryLifecycle.test.ts` adds `arms when the user cancels during pre-query work`, which holds the pre-query callback pending, cancels through a real `QueryGuard`, and verifies the reminder is armed.

### M2. Vacuous session-isolation assertion

**Fixed** in `236a90c0`. The lifecycle test now starts and cancels a second model-bound query before changing from session A to session B, so the mismatched-session assertion runs with pending state instead of after the prior reminder was consumed. The pure reminder test continues to assert that a mismatched session consumes the pending state without producing a reminder.

### M3. Implicit `workload` / `agentId` invariants

**Fixed** in `236a90c0`. `isNormalLocalUserPrompt` now requires both fields to be absent, making the human-only consumption boundary explicit. The prompt-classification test covers both programmatic provenance fields while retaining the positive keyboard-prompt case.

### L1. Consumption before `processUserInput`

**Not changed.** The intended contract is to clear the one-shot after the next normal prompt whether or not prompt processing or the query succeeds. Moving consumption after `processUserInput` would preserve stale correction state after a failed normal submission and contradict that contract.

### L2. Non-prompt submit interruption

**Not changed.** The interrupt-on-submit branch is deliberately limited to non-prompt input and does not route through the Escape/user-cancel lifecycle. Slash and local UI commands neither receive nor consume correction context; the next normal prompt remains the only eligible consumer.

### L3. `mergeUserMessages` meta handling

**Not changed.** In the normal path, `mergeUserMessages` spreads the first message and therefore preserves its `isMeta` field. When history snipping is enabled, the explicit rule at `src/utils/messages.ts:1868-1884` clears `isMeta` if either merged operand is real user content, while the reminder text remains model-visible; this is existing intentional behavior rather than loss introduced by this PR.

### L4. Transcript persistence and resume

**Not changed.** Persisting hidden/meta messages follows the existing transcript convention requested for this change. SDK/internal mapping preserves the synthetic/meta marker (`src/utils/messages/mappers.ts`), and the reminder remains in its original chronological position adjacent to the correction rather than being re-injected on resume.

### L5. Session-mismatch consumption

**Not changed.** Dropping a pending reminder when the active session changes is the session-isolation boundary: deferring it would allow correction context created in one session to appear later after switching away and back. The mismatch behavior is covered directly in `interruptionCorrection.test.ts` and through the re-armed lifecycle case.

### L6. Re-reading the active guard during cancellation

**Not changed.** `handleCancellation` re-reads `QueryGuard.activeContext` synchronously before `forceEnd`; there is no await or other interleaving between the captured context and that read. The tracker also compares the active query with the model-bound query before arming or clearing, so no current failure path is present.

### L7. Query ID sources

**Not changed.** The tracker does not independently trust the caller ID: binding requires the supplied query ID to equal `QueryGuard.activeContext.queryId`, and cancellation reads the guard again before comparing it with the stored binding. Those checks make the guard the source of truth on both transitions.

Validation observed on `236a90c0`:

- `bun test src/screens/REPL.queryLifecycle.test.ts src/utils/interruptionCorrection.test.ts src/utils/handlePromptSubmit.test.ts src/utils/profilerRetention.test.ts src/utils/messages/apiTransform.test.ts src/utils/messages/factories.test.ts` — 41 passed, 0 failed
- `bun run typecheck` — passed
- `bun run build` — passed
- `bun run smoke` — passed (`0.23.0`)
- `bun run security:pr-scan` — no suspicious additions
- `git diff --check` — clean

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Keep Remote Control turns from arming the local correction reminder  
  `src/screens/REPL.tsx:2323`
  Remote Control inbound messages are enqueued with `bridgeOrigin: true` in `useReplBridge`, but execute through the local `QueryGuard`; they do not set `activeRemote.isRemoteMode`. If a local user presses Escape while one of those bridge-origin turns is active, this call passes `isRemoteMode: false` and arms the tracker. The next normal local prompt then receives a reminder claiming it is a correction to the remote turn, despite the PR's stated exclusion for remote events. Preserve the bound turn's bridge/provenance eligibility (or otherwise reject bridge-origin turns when arming) so only a locally submitted model turn can create the reminder.

**Review by jatmn — CHANGES_REQUESTED** [review]
## Findings

### P1 — Auto-restore after early Esc still arms “correction” framing

**Files:**  
- `src/screens/REPL.tsx` (~2323–2336 arm; ~3308–3332 auto-restore; ~4015–4021 restore)  
- `src/utils/interruptionCorrection.ts` (~101–118, ~127–134)  
- `src/utils/handlePromptSubmit.ts` (~515–543)

**Problem:** Escape of a model-bound turn arms `pendingSessionId`. When Esc happens before meaningful assistant content, `onQuery`’s finally block **auto-restores** the interrupted user prompt (rewind history + put the same text back into the input). That path never disarms the tracker.

Resubmitting the restored (often unchanged) prompt then consumes the one-shot and injects:

> Treat the user's latest message as a correction and do not continue the interrupted plan…

But auto-restore already removed the interrupted turn from the conversation. Product-wise this is a **redo**, not a correction of a surviving plan — and the reminder steers the model the wrong way.

**Repro:**

1. Start a normal local multi-step prompt that begins a model turn.  
2. Press Esc immediately (auto-restore conditions: `user-cancel`, idle guard, empty input, empty queue, synthetic-only tail).  
3. Observe the original prompt restored into the input and transcript rewound.  
4. Press Enter without changing the text.  
5. Observe the meta interruption-correction reminder prepended to `newMessages` for the resubmit.

**Expected:** auto-restore should clear pending (or arming should not apply when the cancel only leaves a synthetic tail that will be rewound).  
**Actual:** pending survives; next normal local submit is framed as a correction.

**Fix direction:** On successful auto-restore, clear tracker pending (dedicated `reset`/disarm). Or only arm when auto-restore will **not** run (e.g. partial assistant content already present / non-synthetic tail). Add a lifecycle test: cancel → auto-restore path → `takeReminder()` is null.

---

### P2 — Destructive `takeReminder()` before successful process / model ownership

**Files:**  
- `src/utils/handlePromptSubmit.ts` (~515–543, ~588–622, ~651–657)  
- `src/utils/interruptionCorrection.ts` (~27–39, ~127–134)  
- `src/utils/processUserInput/processUserInput.ts` (~197–227)

**Problem:** For a normal local prompt, `takeInterruptionCorrectionReminder()` runs **before** `await processUserInput(...)` and always clears `pendingSessionId`. Failures after take permanently burn the one-shot:

1. **`processUserInput` throws** (attachments / hooks / unexpected errors) — reminder never pushed; pending already null.  
2. **`shouldQuery: false` after take** (e.g. UserPromptSubmit block / `preventContinuation`) — reminder may be appended to transcript noise while no model correction turn runs; next real correction has no framing.  
3. **Concurrent `onQuery` decline** (`tryStart === null` at `REPL.tsx` ~3110–3125) — meta reminder is filtered out of re-enqueue (`!m.isMeta`); pending already cleared; re-enqueued command also lacks `preExpansionValue`, so it is no longer `isNormalLocalUserPrompt`.

Author comment L1 intentionally clears on the next normal *attempt* even if the query fails. That is a reasonable policy for “don’t leave stale correction state,” but it still creates silent **loss of the feature payload** on throw and concurrent-reenqueue paths that never deliver the reminder to a model turn. Tests only cover happy-path prepend + slash preserve.

**Fix direction (pick one coherent contract):**

- **Consume-on-successful-delivery:** take only after `processUserInput` succeeds with `shouldQuery === true`, and only after `tryStart` owns the turn (or re-arm when `onQuery` returns `false`).  
- **Keep consume-on-attempt:** still restore pending if the reminder was never pushed, or if concurrent re-enqueue drops it; re-enqueue with `preExpansionValue` and preserve/re-arm meta.

Add tests for throw-after-take, `shouldQuery: false` after take, and concurrent re-enqueue.

---

### P2 — Queued mid-turn follow-up is labeled a “correction” after Esc

**Files:**  
- `src/utils/handlePromptSubmit.ts` (~344–380 enqueue; ~515–543 take)  
- `src/screens/REPL.tsx` (~3317–3322 auto-restore skipped when queue non-empty)

**Problem:** While turn A is loading, the user can queue B. Esc A arms pending; auto-restore is skipped when the queue is non-empty. Queue processor then runs B as a normal local prompt and injects “treat latest message as a correction / do not continue the interrupted plan.”

B was often typed as a **planned next step**, not a reaction to the interrupt. Framing it as a correction of A is product-skewed.

**Repro:** long turn A → submit B (queued) → Esc A → B runs with interruption-correction reminder.

**Fix direction:** Do not arm when a human prompt is already queued; or only inject for prompts submitted **after** the interrupt (generation / timestamp); or soften reminder text for queued-before-cancel cases.

---

### P3 — Default `mergeUserMessages` can keep `isMeta` from the reminder when HISTORY_SNIP is off

**Files:**  
- `src/utils/messages.ts` (~1868–1906)  
- `src/utils/handlePromptSubmit.ts` (~541–560)  
- `src/utils/interruptionCorrection.ts` (~34–37)

**Problem:** Injection always produces consecutive user messages `[meta reminder, real user]`. `normalizeMessagesForAPI` merges them. HISTORY_SNIP runtime correctly sets `isMeta: a.isMeta && b.isMeta`; the default path spreads `...a` and can leave **`isMeta: true` on the merged turn that includes real user text**.

UI still renders pre-normalize messages (reminder hidden, user visible), and API content still includes the reminder text, so this is not a full wire drop. Residual risk is any post-normalize consumer that treats `isMeta` as “not a real user turn” (routing helpers, snip tagging, future callers). This PR makes the meta+real merge the common interrupt path.

**Fix direction:** Always use AND semantics for `isMeta` on merge (ungated), matching the HISTORY_SNIP comment. Add a normalize unit test for `[meta reminder, real user]`.

---

**Comment by chioarub** [comment]
Thanks for the detailed review. The four lifecycle and normalization findings are addressed in `64393c26`, with consume-on-attempt retained for prompts that actually begin submission processing.

### Auto-restore retained stale correction framing

**Fixed** in `64393c26`. Conversation rewrites now clear the pending reminder at the shared rewind boundary after the transcript is replaced (`src/screens/REPL.tsx:3973-3974`), so early-Escape auto-restore and manual conversation restore cannot frame a redo as a correction. Successful partial-summary context replacement clears the same state at `src/screens/REPL.tsx:5232-5240`; code-only restore does not alter it. `clears a pending reminder when the interrupted context is rewritten` covers the tracker lifecycle.

### Reminder consumption before model ownership

**Fixed with consume-on-attempt preserved** in `64393c26`. The issue contract requires the one-shot to clear after the next normal prompt even when processing fails, so a `processUserInput` exception still consumes it; `consumes a reminder when processing the normal prompt throws` pins that behavior. A blocked `shouldQuery: false` prompt no longer appends the reminder to transcript messages, while a concurrent guard rejection now restores the pending reminder and requeues the same prompt with its local provenance intact (`src/utils/handlePromptSubmit.ts:508-513`, `:560-566`, and `:647-651`). The blocked-prompt, concurrent-rearm, and provenance tests cover those branches.

### Prompt queued before Escape was treated as corrective

**Fixed** in `64393c26`. Cancellation now checks the live command queue and does not arm when a normal local human prompt is already queued (`src/screens/REPL.tsx:2325-2330`; `src/utils/interruptionCorrection.ts:21-27`). Remote, slash, meta, workload, and programmatic submissions remain ineligible through the existing normal-local prompt predicate.

### Meta reminder merged with real user input remained meta

**Fixed** in `64393c26`. `mergeUserMessages` now applies AND semantics to `isMeta` unconditionally, so any real user operand makes the combined API turn non-meta (`src/utils/messages.ts:1868-1873`). `normalizing meta context with real user input produces a non-meta message` covers the API normalization boundary, and the context-collapse tests verify that `isCollapseSummary` independently keeps collapsed summaries non-snippable.

Validation:

- `bun test src/screens/REPL.queryLifecycle.test.ts src/utils/interruptionCorrection.test.ts src/utils/handlePromptSubmit.test.ts src/utils/messages/apiTransform.test.ts src/services/contextCollapse/operations.test.ts src/utils/profilerRetention.test.ts src/utils/messages/factories.test.ts` — 59 passed, 0 failed
- `bun run typecheck` — passed
- `bun run build` — passed
- `bun run security:pr-scan` — passed
- `git diff --check` — passed
- `bun run check` — build, smoke, and deadcode passed; the full test stage reproduced the documented baseline of 6805 passed, 2 skipped, and the same 43 unrelated failures

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Keep the one-shot correction reminder out of later API requests
  `src/screens/REPL.tsx:3145`
  The reminder is inserted into `newMessages` and then permanently appended to the REPL history here. Every later turn normalizes and sends that full history, so after an interruption followed by `do Y instead`, an unrelated later prompt still receives the old `<system-reminder>` telling the model to treat the latest message as a correction and not continue the interrupted plan. This also survives resume because meta messages are persisted. Deliver the reminder only with the correction request (or remove it from durable history immediately afterward) so its stated one-shot lifetime is actually enforced.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Clear pending correction context when a full compaction rewrites the conversation
  `src/screens/REPL.tsx:2857`
  A pending reminder is intentionally cleared for rewinds and partial compactions, but this full compact-boundary path replaces the active conversation without clearing it. Repro: cancel an eligible turn, run `/compact` (which does not consume the reminder), then submit an unrelated normal prompt. The session ID is unchanged, so `takeReminder()` injects “The previous assistant turn was interrupted” into the new compacted conversation even though that turn is no longer present. Clear the tracker when applying a successful full compact boundary, and cover the cancel → `/compact` → prompt sequence.

**Comment by chioarub** [comment]
Thanks for the follow-up. The full-compaction lifecycle gap is fixed, and I also rechecked the adjacent request-only failure paths raised during review.

### Pending correction context survives a full compaction

**Fixed** in `92c07fff`. The REPL's synchronous message-state transition now detects when a new latest compact-boundary UUID is applied and clears pending correction state before replacing the message ref. Manual, streamed full/auto, and partial compaction all use that transition; `clears a pending reminder when a full compact boundary is applied` and `clears a pending reminder when a later compaction replaces an existing boundary` cover first and repeated compactions, including the functional updater used by fullscreen compaction.

### Restore the reminder when preprocessing or query execution throws

**Not changed.** The one-shot is intentionally consumed by the next normal prompt attempt, even if that attempt later fails; `consumes a reminder when processing the normal prompt throws` pins that behavior. Restoring after an attempted correction could attach stale correction framing to a later unrelated prompt. I would change this if the contract moves from consume-on-attempt to consume-on-successful-model-delivery semantics.

### Add a no-human-turn fallback test for request-only insertion

**Not changed.** The REPL supplies request-only correction context only alongside a normal local human prompt, so the feature path does not enter the no-human fallback. The fallback remains defensive for other `QueryParams` callers; I would add direct coverage if a supported caller begins supplying request-only messages without a human turn.

Validation:

- `bun test src/utils/interruptionCorrection.test.ts src/screens/REPL.queryLifecycle.test.ts src/utils/handlePromptSubmit.test.ts src/query/requestOnlyMessages.test.ts` — 26 passed, 0 failed
- `bun run typecheck` — passed
- `bun run smoke` — passed
- `bun run security:pr-scan` — passed
- `git diff --check` — passed

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Preserve the non-snip meta contract when merging unrelated messages
  `src/utils/messages.ts:1869`
  This removes the `HISTORY_SNIP` runtime guard and clears `isMeta` for every adjacent meta/real-user merge, not just the new request-only reminder. In builds without HISTORY_SNIP, a meta image/PDF attachment merged with its user prompt now becomes non-meta. If that attachment triggers one of the synthetic image/PDF/request-size API errors, `normalizeMessagesForAPI()` no longer recognizes it as the preceding meta message and therefore does not strip the rejected block on the next request (`src/utils/messages.ts:1443-1565`). The same invalid attachment is resent and the session remains stuck on the recurring provider error. Keep the existing non-snip merge semantics (or limit the new behavior to the interruption reminder) and add a recovery regression test.

- [P2] Do not discard an earlier correction reminder when auto-restoring a later aborted correction
  `src/screens/REPL.tsx:3977`
  Interrupt turn A after it has produced output, submit correction B, then press Escape while B is still in pre-query work. B consumes A's request-only reminder, but the second interruption auto-restores B because its tail is synthetic-only. `rewindConversationTo()` unconditionally clears the tracker even though it only removes B and leaves interrupted A in the history. Resubmitting B then reaches the model without the correction framing that was never delivered. Preserve or re-arm the prior reminder when an auto-restore keeps its interrupted turn, and cover the A -> B -> Escape -> resubmit sequence.

**Comment by chioarub** [comment]
Thanks for the thorough review. I pushed an additive update that merges the current base and closes the interruption-recovery and attachment-retry gaps.

### Non-`HISTORY_SNIP` meta-message merging

**Fixed** in `257ae5f5`. `mergeUserMessages()` now clears `isMeta` only when `HISTORY_SNIP` is compiled in and its runtime path is enabled, preserving the legacy meta merge used by attachment API-error recovery otherwise. `messages/apiTransform.test.ts` covers both feature modes and verifies that a rejected oversized image is stripped while the user's prompt is retained.

### Correction reminder lost after interrupting the corrective turn

**Fixed** in `257ae5f5`. Auto-restore now rewinds through the correction tracker while preserving the earlier request-only reminder only when that reminder was actually present in the interrupted request. `interruptionCorrection.test.ts` covers A -> correction B -> Escape during B pre-query -> auto-restore -> resubmit, including exactly-once consumption.

The follow-up in `f495c6cd` also clears stale correction state when a compact boundary is removed and pins programmatic edit/restore cancellation so it cannot re-arm an already consumed reminder.

### Runtime-off `HISTORY_SNIP` test scenario

**Not changed.** The current `isSnipRuntimeEnabled()` contract is an invariant `true`, so a build with `HISTORY_SNIP` enabled and that helper returning `false` is not a reachable production state. Importing the runtime helper directly into the message test also introduces a circular module load under the feature build; the reachable feature-off and feature-on paths are covered instead. If the runtime helper becomes configurable, that new contract would warrant a dedicated third branch test.

Validation on `259336d8`:

- focused REPL/message/compaction/status-line tests: 72 passed, 0 failed
- `HISTORY_SNIP` focused tests: 21 passed, 1 intentional non-snip skip, 0 failed
- `bun run typecheck`: passed
- `bun run build`: passed
- `bun run security:pr-scan`: passed
- `git diff --check`: passed

Hosted Node 22, typecheck, and web checks passed. The Node 24.11 full-suite job reproduced the existing broad shared-process baseline (6,906 passed, 2 skipped, 43 unrelated failures); all PR-specific suites passed in that job. The fork account does not have permission to rerun that failed upstream job.

File: src/screens/REPL.queryLifecycle.test.ts [hunk 1]
```diff
@@ -53,4 +55,151 @@ describe('REPL query lifecycle timeout logging', () => {
...
+    expect(tracker.takeReminder()).toBeNull()
+  })
+
+  test('keeps maxTurns in the query callback dependencies', () => {
+    const start = source.indexOf('const onQueryImpl = useCallback')
+    expect(start).toBeGreaterThan(-1)
+    const end = source.indexOf('const onQuery = useCallback', start)
+    expect(end).toBeGreaterThan(start)
+    expect(source.slice(start, end)).toContain('titleDisabled, maxTurns,')
+  })
```
[the remark this answers is not in the document]
Comment (jatmn) [reply src/screens/REPL.queryLifecycle.test.ts]: Addressed in 7900a53c: the implementation-detail `source.indexOf` assertion was removed. The runtime behavior is unchanged by this test-only cleanup; coverage for max-turn resolution and query parameter propagation remains in `replMaxTurnsProp.test.ts`.

**Review by jatmn — APPROVED** [review]
@kevincodex1 LGTM

**Review by kevincodex1 — APPROVED** [review]
